digdag-standards/src/main/java/io/digdag/standards/operator/aws/EmrOperatorFactory.java

```diff
@@ -0,0 +1,1753 @@
+                }
+                catch (RetryExecutor.RetryGiveupException e) {
+                    logger.info("Failed to delete staging files in {}", staging.get(), e.getCause());
+                }
+
+                req.setMarker(res.getMarker());
+            }
+            while (req.getMarker() != null);
+        }
+    }
+
+    private static class StepCompiler
+    {
+        private final ParameterCompiler pc;
+        private final String tag;
+        private final List<Config> steps;
+        private final Filer filer;
+        private final ObjectMapper objectMapper;
+        private final String defaultActionOnFailure;
+
+        private List<StepConfig> configs;
+
+        private int index;
+        private Config step;
+        private RemoteFile runner;
+
+        StepCompiler(String tag, List<Config> steps, Filer filer, ParameterCompiler pc, ObjectMapper objectMapper, String defaultActionOnFailure)
+        {
+            this.tag = Preconditions.checkNotNull(tag, "tag");
+            this.steps = Preconditions.checkNotNull(steps, "steps");
+            this.filer = Preconditions.checkNotNull(filer, "filer");
+            this.pc = Preconditions.checkNotNull(pc, "pc");
+            this.objectMapper = Preconditions.checkNotNull(objectMapper, "objectMapper");
+            this.defaultActionOnFailure = Preconditions.checkNotNull(defaultActionOnFailure, "defaultActionOnFailure");
+            Preconditions.checkArgument(!steps.isEmpty(), "steps");
+        }
+
+        private void compile(RemoteFile runner)
+                throws IOException
+        {
+            this.runner = runner;
+            configs = new ArrayList<>();
+            index = 1;
+            for (int i = 0; i < steps.size(); i++, index++) {
+                step = steps.get(i);
+                String type = step.get("type", String.class);
+                switch (type) {
+                    case "flink":
+                        flinkStep();
+                        break;
+                    case "hive":
+                        hiveStep();
+                        break;
+                    case "spark":
+                        sparkStep();
+                        break;
+                    case "spark-sql":
+                        sparkSqlStep();
+                        break;
+                    case "script":
+                        scriptStep();
+                        break;
+                    case "command":
+                        commandStep();
+                        break;
+                    default:
+                        throw new ConfigException("Unsupported step type: " + type);
+                }
+            }
+        }
+
+        List<StepConfig> stepConfigs()
+        {
+            Preconditions.checkState(configs != null);
+            return configs;
+        }
+
+        private String localWorkingDirectory()
+        {
+            return LOCAL_STAGING_DIR + "/" + tag + "/steps/" + index + "/wd";
```

Comment:
[minor] Maybe passing `index` as an argument and removing `index` instance variable would be better in terms of maintainability